pkg/imageverifycache/client.go

```diff
@@ -0,0 +1,120 @@
+package imageverifycache
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/ristretto"
+	"github.com/go-logr/logr"
+	v1 "github.com/kyverno/kyverno/api/kyverno/v1"
+)
+
+type cache struct {
+	logger         logr.Logger
+	isCacheEnabled bool
+	maxSize        int64
+	ttl            time.Duration
+	lock           sync.Mutex
+	Cache          *ristretto.Cache
+}
+
+type Option = func(*cache) error
+
+func New(options ...Option) (Client, error) {
+	cache := &cache{}
+	for _, opt := range options {
+		if err := opt(cache); err != nil {
+			return nil, err
+		}
+	}
+	config := ristretto.Config{
+		MaxCost:     cache.maxSize,
+		NumCounters: 10 * cache.maxSize,
+	}
+	rcache, err := ristretto.NewCache(&config)
+	if err != nil {
+		return nil, err
+	}
+	cache.Cache = rcache
+	return cache, nil
+}
+
+func DisabledImageVerifyCache() Client {
+	return &cache{
+		logger:         logr.Discard(),
+		isCacheEnabled: false,
+		maxSize:        0,
+		ttl:            0,
+	}
+}
+
+func WithLogger(l logr.Logger) Option {
+	return func(c *cache) error {
+		c.logger = l
+		return nil
+	}
+}
+
+func WithCacheEnableFlag(b bool) Option {
+	return func(c *cache) error {
+		c.isCacheEnabled = b
+		return nil
+	}
+}
+
+func WithMaxSize(s int64) Option {
+	return func(c *cache) error {
+		c.maxSize = s
+		return nil
+	}
+}
+
+func WithTTLDuration(t time.Duration) Option {
+	return func(c *cache) error {
+		c.ttl = t
+		return nil
+	}
+}
+
+func (c *cache) Set(ctx context.Context, policy v1.PolicyInterface, ruleName string, imageRef string) (bool, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.logger.Info("Setting cache", "policy", policy.GetName(), "ruleName", ruleName, "imageRef", imageRef)
+	if !c.isCacheEnabled {
+		return false, nil
+	}
+	key := string(policy.GetUID()) + ";" + policy.GetName() + ";" + ruleName + ";" + imageRef
+	if c.ttl <= 0 {
+		stored := c.Cache.Set(key, nil, 1)
+		if stored {
+			c.logger.Info("Successfully set cache", "policy", policy.GetName(), "ruleName", ruleName, "imageRef", imageRef)
+			return true, nil
+		}
+		return false, nil
+	}
```

Comment:
Instead of this, add some default values for the TTL in the struct when we initialize it as we rely on TTL for clearance
1hr seems fine to me
cc @eddycharly

Reply:
done